[spirv] Add support for [[vk::set(X)]]

pow2clk opened the pull request:
A new attribute [[vk::set(X)]] can be used to specify the descriptor
set of a resource variable, leaving the binding index to be
determined automatically based on what's available.

Fixes https://github.com/Microsoft/DirectXShaderCompiler/issues/1340

pow2clk commented:
Probably going to need input from @antiagainst and/or @ehsannas

ehsannas commented on tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:
```diff
@@ -1096,8 +1100,9 @@ class BindingShiftMapper {
 bool DeclResultIdMapper::decorateResourceBindings() {
   // For normal resource, we support 3 approaches of setting binding numbers:
```
s/3 approaches/4 approaches

ehsannas commented on tools/clang/lib/SPIRV/DeclResultIdMapper.cpp:
```diff
@@ -1110,7 +1115,8 @@ bool DeclResultIdMapper::decorateResourceBindings() {
   // In the following order:
```
Need to update the comment above:
```
   // In combination, we need to handle 12 cases:
   // - 4 cases for nomral resoures (m1, m2, m3, m4)
   // - 8 cases for associated counters (mX * cY)
```

pow2clk commented:
I inferred from the fact that there is a priority order rather than an error when both vk:binding and the register annotations are included that it was the preferred method even though not producing an error or at least a warning in that case seemed to me like a mistake. If we are to add an error for ambiguously over-specifying a resource one way, should we do the same for the other way?

ehsannas commented:
hmm... yeah, you're right. The priority order would make the error redundant. I remember now: we used to have warnings for these, but it turned out to be cause a large set of warnings on big workloads.

pow2clk commented:
Well, I'm content to add warnings for all such cases, add a warning for only this new case, or add no warnings. If you'd like to mull it over or consult on the issue, I don't think decision on this issue is on the critical path toward submitting the change.

ehsannas commented:
Yeah, let's not introduce any warnings/errors in this CL (to be consistent with the current behavior).

pow2clk commented:
I believe all the changes are in place. I added to the test you highlighted and added another of my own meant to exhaustively test the more complicated precedence operations. This was inspired by a bug I found by inspection where explicitly defined counter bindings would end up in the wrong descriptor set (even a different one from the resource) because the register annotation would take precedence over any other setting. The new test covers that and a lot more.

pow2clk commented:
Only now does it occur to me that specifying both the `binding` only and the `set` might be most naturally expected to specify the individual binding and the individual set. i.e. [[vk::set(1), vk::binding(3)]] will currently use the default descriptor set 0 with binding 3 because `binding` trumps `set` while it would be more intuitive to interpret it as descriptor set 1 with binding 3. 

The drawbacks of implementing it that way is that presumably, we'd want a fully specified `vk::binding(3,2)` to ignore `vk::set(1)`. Which complicates and confuses the issue somewhat.

I would posit that the least confusing solution would be to have `binding` only take one argument and only specify binding and provide `set` to specify the set exclusively. If one is present and the other is missing, the missing attribute will be implicit. That would break backward compatibility of course, but it's the cleanest solution to my thinking.

ehsannas commented:
@pow2clk It looks like you have accidentally pushed several linux-port commits into this pull-request.

ehsannas commented:
It occurred to me too. Using something like `[[vk::set(1), vk::binding(3)]]` looks less complicated and more natural to me. But I'm afraid there are people who rely on the old syntax (`binding(x,y)`). If we want to deprecate the old syntax, we'd need some more time to discuss it.

pow2clk commented:
Hmm. Not sure how that happened. Guess it was past my bedtime. I'll -f it all away.

I'll keep things as they are for now per your advice.

pow2clk commented:
Okay. I'm clearly doing something wrong here. My branch is tracking Microsoft master and only has my change, but it's pushing all kinds of other changes. Gimme a minute.

pow2clk commented:
@ehsannas I believe I have responded to your changes. Can you confirm?

ehsannas commented:
+1.
Thanks @pow2clk ! Yes, I'd like to get a +1 from @antiagainst before merging.

pow2clk commented:
@ehsannas and I discussed generating an error or warning earlier. The only reason against it was consistency with earlier behavior regarding `vk::binding` versus `:register`. Personally, I think it makes sense to throw an error or warning whenever duplicate specifications are combined, but I was told that would clash with a lot of existing code. I don't know if that's tests included here or code out in the wild. It seems that if you use any of these three with any other, you're probably not doing it on purpose and might be surprised by the results.

That said, I remain willing to produce warnings or errors for all, any, or none of the potential combinations even if it means changing tests.

antiagainst commented:
IMHO, they are slightly different. Let me unroll this issue a little bit.

### Guidelines

We have a few guidelines when designing language features:

1. Leverage existing langues constructs first. Then attribute, then builtin types.
2. The more specific a feature is, the more priority it gets.
3. Portability is important.
4. The feature should be intuitive, consistent, and easy to use.
5. Provide meaningful defaults, together with detailed control mechanisms.

With the above in mind, let's look at this descriptor set assignment issue.

### Descriptor assignment

Looking at Rule No.1. HLSL already have native constructs: `:register`. Optimally we should use it, but sadly it cannot map 1:1 to Vulkan semantics. `:register(xX, spaceY)` has three parameters: `x`, `X`, and `Y`, while Vulkan only has set # and binding #, leaving register type `x` mapping to nothing. But we still want to make `:register` work because of Rule No.3, so that ppl with their HLSL code initially targeting DX  (esp. via `fxc.exe`) aren't forced to do modifications. Therefore, we have `-fvk-*-shift` command-line options.

But that's not the whole story. We also like an explicit way to control the descriptor assignment (Rule No.5) for ppl that are willing to put `#ifdef`s and inject all the binding facts in their source code. So we have `[[vk::binding]]`.

Because of Rule No.2, `[[vk::binding]]` overrides `:register` since `[[vk::binding]]` is more specific (only for Vulkan) while `:register` is more general (for both DX and VK). We could define it as an error by using both `[[vk::binding]]` and `:register`, but that would mean ppl need to `#ifdef` away `:register`s. We think that's annoying for devs. Arguably, VK specific constructs to override general DX+VK constructs is anstraightforward rule for devs to grasp implicitly in their mind. So we didn't set it as an error. VK specific construct to override native construct is a consistent rule (a derive of Rule No.2) we have for all cases I believe.

Then about `[[vk::set]]`, which is also a VK specific construct to be introduced. It is not "equivalent" to `:register` when considering overriding behavior. Rather, it's more at the same level as `[[vk::binding]]`. One could still argue that since `[[vk::binding]]` is more specific, so it should override `[[vk::set]]`. But at the native-vs-external/general-vs-specific aspect, it is not. And that's the aspect I think which is more important.

Besides, having `[[vk::binding]]` overriding `[[vk::set]]` is gonna be a source of confusion as said before, since it is more natural to read `[[vk::set(...), vk::binding(...)]]` as a combination for setting set and binding separately (sadly, `[[vk::binding]]` was not designed like that). So, instead of overriding and give surprises to devs, emitting an error would be more clear and simplifies the impl. too. And it won't hurt the use cases: if you want only specify set #, use `[[vk::set]]`; if you want both set # and binding #, use `[[vk::binding]]`. If you want only binding # but don't care set number, well I feel that's very weird. ;-P

antiagainst commented:
With the above said, truly thanks for this pull request. It's great: well written and tested. But as said in #1340, right now we are striving for needs based changes. So we may keep this one in our back pocket and merge it when needed.

pow2clk commented:
I'm in no position to argue the need for this change. I just implemented it. FWIW I don't see the discussion on #1340 as having come to any conclusions. I've asked @nsubtil to comment since he asked for it.

In the meantime I'll pare this down to the bug it fixes and what remains of the test case without vk::set and submit it as a separate PR.

nsubtil commented:
I don't really understand the "needs based changes" argument. I am building Vulkan software and filed this feature request because I have a need for it.

The current setup actively discourages developers from separating bindings into multiple sets, as it requires manually assigned locations for every set except for set 0. Larger apps tend to use reflection to pull out binding locations and the compiler already has a mechanism to automatically assign locations, which makes this setup convenient --- but it effectively only seems to work on set 0. If this was not an intended usage pattern, then I question why we have automatic location assignments in the first place.

If the conclusion is that there is a design flaw in the current syntax for binding location assignment, then perhaps that should be fixed as well?

pow2clk commented:
In case a supplicant should wander in from the desert, request this feature, and be deemed worthy and untrammeled by tainted associations, I wanted to have this in the best state for such a fortuitous circumstance.

So here I've added errors where vk::set is used with vk::binding, :register now takes precedence over vk::set, the precedence test is updated for these new relations and another is added for the new errors.

chaoticbob commented:
> I don't really understand the "needs based changes" argument. I am building Vulkan software and filed this feature request because I have a need for it.

Understood that you're asking for for this. Adding this feature is a non-trivial change to one aspect of the workflow in specifying set/binding data. We would like to add this request when there is more feedback from the community on its need. 

There are a few potential gaps with this solution that we'd like to better understand how to handle before adding it:

**1. What's the expected output of this case:**
```
Texture2D MyTexA : register(t0, space2);
Texture2D MyTexF : register(t2, space2);
[[vk::set(2)]] Texture2D MyTexB;
[[vk::set(2)]] Texture2D MyTexC;
[[vk::set(2)]] Texture2D MyTexD;
[[vk::set(2)]] Texture2D MyTexE;
```

**2. Are developers expected to know about potential misalignments?**

VertexShaderSource.hlsl
```
[[vk::set(2)]] Texture2D MyTexB; // binding = 0
[[vk::set(2)]] Texture2D MyTexC; // binding = 1
[[vk::set(2)]] Texture2D MyTexD; // binding = 2
[[vk::set(2)]] Texture2D MyTexE;  // binding = 3
```
PixelShaderSource.hlsl
```
[[vk::set(2)]] Texture2D MyTexB; // binding = 0
[[vk::set(2)]] Texture2D MyTexC; // binding = 1
[[vk::set(2)]] Texture2D MyTexE;  // binding = 2
```
In this case the resource MyTexE will share the same name between the who modules, but the binding will be different.

**2. Is there a potential hazard with descriptor indexing?**
```
[[vk::set(2)]] Texture2D MyTexB; // binding = 0
[[vk::set(2)]] Texture2D MyTexC; // binding = 1
[[vk::set(2)]] Texture2D MyTexD; // binding = 2
[[vk::set(2)]] Texture2D MyTexE[];  // binding = 3
Texture2D MyTexF[] : register(t3, space2);
```
Who gets to claim the last entry in the set for this case?

> The current setup actively discourages developers from separating bindings into multiple sets, as it requires manually assigned locations for every set except for set 0. Larger apps tend to use reflection to pull out binding locations and the compiler already has a mechanism to automatically assign locations, which makes this setup convenient --- but it effectively only seems to work on set 0. If this was not an intended usage pattern, then I question why we have automatic location assignments in the first place.

Come on, this is just not true. At most one could argue that the there is a missing convenience factor without something like `[[vk:::set]]`. The usage pattern for this is derived from FXC's behavior with a mix of what glslang brought to the table. AFAIK, there isn't a parallel version of this in HLSL/DirectX. 

> If the conclusion is that there is a design flaw in the current syntax for binding location assignment, then perhaps that should be fixed as well?

We have not had any feedback from ISVs that this has been a problem. If you're aware of any that needs to be addressed, we're more than happy to consider the discussion.

**FWIW**: It sounds like what you may be after is the ability to have a more flexible system to specify the root signature in the HLSL source.

nsubtil commented:
I may be missing something, but most of the issues you raised already have to be handled in set 0. My answer is generally to implement the same behavior that's already in place for set 0 (or in other words: auto-assign locations for each set independently), so I'm not sure how useful my feedback is --- I may just need a more detailed explanation of the problem?

> 1. What's the expected output of this case?
> 
> Texture2D MyTexA : register(t0, space2);
> Texture2D MyTexF : register(t2, space2);
> [[vk::set(2)]] Texture2D MyTexB;
> [[vk::set(2)]] Texture2D MyTexC;
> [[vk::set(2)]] Texture2D MyTexD;
> [[vk::set(2)]] Texture2D MyTexE;

I'd expect the same behavior that the compiler implements today if you remove all the space and vk::set decorations. This doesn't look like a new problem.

> 2. Are developers expected to know about potential misalignments?
> 
> VertexShaderSource.hlsl
> 
> [[vk::set(2)]] Texture2D MyTexB; // binding = 0
> [[vk::set(2)]] Texture2D MyTexC; // binding = 1
> [[vk::set(2)]] Texture2D MyTexD; // binding = 2
> [[vk::set(2)]] Texture2D MyTexE;  // binding = 3
>
> PixelShaderSource.hlsl
> 
> [[vk::set(2)]] Texture2D MyTexB; // binding = 0
> [[vk::set(2)]] Texture2D MyTexC; // binding = 1
> [[vk::set(2)]] Texture2D MyTexE;  // binding = 2
> In this case the resource MyTexE will share the same name between the who modules, but the binding will be different.

Same answer as above. Doesn't seem like a new problem --- if you remove all the vk::set decorations, you get the same issue but in set 0. I'd expect to see whatever behavior we implement today for set 0 apply in other sets.

> 3. Is there a potential hazard with descriptor indexing?
> 
> [[vk::set(2)]] Texture2D MyTexB; // binding = 0
> [[vk::set(2)]] Texture2D MyTexC; // binding = 1
> [[vk::set(2)]] Texture2D MyTexD; // binding = 2
> [[vk::set(2)]] Texture2D MyTexE[];  // binding = 3
> Texture2D MyTexF[] : register(t3, space2);
> Who gets to claim the last entry in the set for this case?

I'm not as familiar with DI, but it seems like this applies without set/space decorators as well?

> 4. How does namespacing for resource types get handled?
> 
> [[vk::set(2)]] Texture2D MyTexB;
> [[vk::set(2)]] RWTexture2D MyRWTexB;
> [[vk::set(2)]] ConstantBuffer MyConstantsB;
> [[vk::set(2)]] Texture2D MyTexC;
> [[vk::set(2)]] RWTexture2D MyRWTexC;
> [[vk::set(2)]] ConstantBuffer MyConstantsC;
> [[vk::set(2)]] Texture2D MyTexD;
> [[vk::set(2)]] RWTexture2D MyRWTexD;
> [[vk::set(2)]] ConstantBuffer MyConstantsD;
> [[vk::set(2)]] Texture2D MyTexE;
> [[vk::set(2)]] RWTexture2D MyRWTexE;
> [[vk::set(2)]] ConstantBuffer MyConstantsE;

Again, doesn't seem like a new problem. Remove vk::set and you're left with an identical situation, I think. But even then, I'm not sure I see what the problem is here.

> Come on, this is just not true. Based on the conversational feedback that we've gotten about reflection vs not reflection, it's about 50/50.

Well, based on my own work on a handful of different engines, I'd say it's 100/0. We both have limited visibility here, but either way it seems like a sizable chunk of the market.

> At most one could argue that the there is a missing convenience factor without something like [[vk:::set]]. The usage pattern for auto asignment is derived from FXC's behavior with a mix of what glslang brought to the table. AFAIK, there isn't a parallel version of [[vk::set]] in HLSL/DirectX.

DX12 is essentially bindless, so the register/location vs. space/set index distinction is not as critical. For Vulkan, descriptor sets matter a lot more, since the API makes a fundamental assumption that the app will split up descriptors into sets to optimize for update frequency.

I wouldn't be surprised if just mimicking FXC's behavior ends up causing friction here, as it was built for a different binding model.

> The minor note here is that it would have potential to break portability for a non-unique feature.

I guess I'm making the assumption that the binding models between DX12 and Vulkan are different enough that it's okay to diverge somewhat if that adds value. That said, I'm not sure what this would break in practice --- can you elaborate?

> FWIW: It sounds like what you may be after is the ability to have a more flexible system to specify the root signature in the HLSL source.

You mean embedding a descriptor set layout description in the shader? That could be useful, but I'm not sure that's what I'm going for here (reflection pretty much gives you that anyway). I'd be happy to go down that path if that's more amenable than this proposal, though.

On a footnote, I'd like to point out that I filed the feature request since it seems like a good idea to me. It could turn out to be a bad idea and I'm open to that. But I'd like to move away from the argument that we're not a customer in discussing this. If there are technical arguments, let's bring them out first and worry about the governance problem later. I'm sympathetic to that side of the issue as well, but mixing the two tends to lead to bad results.

chaoticbob commented:
> I may be missing something, but most of the issues you raised already have to be handled in set 0. My answer is generally to implement the same behavior that's already in place for set 0 (or in other words: auto-assign locations for each set independently), so I'm not sure how useful my feedback is --- I may just need a more detailed explanation of the problem?

There isn't a problem here to solve. The ask here is to clarify the user end expectations and understand what's in the PR. Should the PR get merged, we'll need to document the functionality and any caveat it may introduce. 

>> 1. What's the expected output of this case?
Texture2D MyTexA : register(t0, space2);
Texture2D MyTexF : register(t2, space2);
[[vk::set(2)]] Texture2D MyTexB;
[[vk::set(2)]] Texture2D MyTexC;
[[vk::set(2)]] Texture2D MyTexD;
[[vk::set(2)]] Texture2D MyTexE;

> I'd expect the same behavior that the compiler implements today if you remove all the space and vk::set decorations. This doesn't look like a new problem.

So to be clear, MyTexB will pick up t1 and MyTexC will pick up t3. You wouldn't expect MyTexB to start at t3, correct?

>> 2. Are developers expected to know about potential misalignments?
VertexShaderSource.hlsl
>> 
>> [[vk::set(2)]] Texture2D MyTexB; // binding = 0
>> [[vk::set(2)]] Texture2D MyTexC; // binding = 1
>> [[vk::set(2)]] Texture2D MyTexD; // binding = 2
>> [[vk::set(2)]] Texture2D MyTexE; // binding = 3
>> 
>> PixelShaderSource.hlsl
>> 
>> [[vk::set(2)]] Texture2D MyTexB; // binding = 0
>> [[vk::set(2)]] Texture2D MyTexC; // binding = 1
>> [[vk::set(2)]] Texture2D MyTexE; // binding = 2
>> In this case the resource MyTexE will share the same name between the who modules, but the >> >> binding will be different.
>> 
>> Same answer as above. Doesn't seem like a new problem --- if you remove all the vk::set decorations, you get the same issue but in set 0. I'd expect to see whatever behavior we implement today for set 0 apply in other sets.

Noted.

>> 3. Is there a potential hazard with descriptor indexing?
>> [[vk::set(2)]] Texture2D MyTexB; // binding = 0
>> [[vk::set(2)]] Texture2D MyTexC; // binding = 1
>> [[vk::set(2)]] Texture2D MyTexD; // binding = 2
>> [[vk::set(2)]] Texture2D MyTexE[]; // binding = 3
>> Texture2D MyTexF[] : register(t3, space2);
>> Who gets to claim the last entry in the set for this case?
>> 
>> I'm not as familiar with DI, but it seems like this applies without set/space decorators as well?

This is likely to generate an error currently. The ask here is to clarify the expectation. For instance, one could argue that since both E and F are the same type and in the same space, they would alias. While it's unlikely to be the behavior, the expectation should be clarified. 

>> 4. How does namespacing for resource types get handled?
>> [[vk::set(2)]] Texture2D MyTexB;
>> [[vk::set(2)]] RWTexture2D MyRWTexB;
>> [[vk::set(2)]] ConstantBuffer MyConstantsB;
>> [[vk::set(2)]] Texture2D MyTexC;
>> [[vk::set(2)]] RWTexture2D MyRWTexC;
>> [[vk::set(2)]] ConstantBuffer MyConstantsC;
>> [[vk::set(2)]] Texture2D MyTexD;
>> [[vk::set(2)]] RWTexture2D MyRWTexD;
>> [[vk::set(2)]] ConstantBuffer MyConstantsD;
>> [[vk::set(2)]] Texture2D MyTexE;
>> [[vk::set(2)]] RWTexture2D MyRWTexE;
>> [[vk::set(2)]] ConstantBuffer MyConstantsE;
>> 
>> Again, doesn't seem like a new problem. Remove vk::set and you're left with an identical situation, I think. But even then, I'm not sure I see what the problem is here.

There's currently two ways this is handled between DXC and glslang. DXC does it based on declaration order. glslang also does declaration order but has a flag that accounts for the namespaces of the resource type (b, s, t, u). 

So the behavior should be purely declaration order?

> I guess I'm making the assumption that the binding models between DX12 and Vulkan are different enough that it's okay to diverge somewhat if that adds value. That said, I'm not sure what this would break in practice --- can you elaborate?

This isn't about the binding model directly. Compiler auto assignment should be orthogonal to that. We strive to maintain source level compatability where we can and only introduce Vulkan/SPIR-V specific attributes when necessary. 

The most I could see is that it adds the possibility for an #ifdef should someone want to do set based auto assignment on the SPIR-V side and something else on the DXIL side.  It would be ideal if there were syntax to do both set/space based auto assignment for both SPIR-V and DXIL. This would extend the language rather than the code-gen. For instance something like:

````
Texture2D MyTexA : register(space3);
Texture2D MyTexF : register(space3);
````

>>FWIW: It sounds like what you may be after is the ability to have a more flexible system to specify the root signature in the HLSL source.

>You mean embedding a descriptor set layout description in the shader? That could be useful, but I'm not sure that's what I'm going for here (reflection pretty much gives you that anyway). I'd be happy to go down that path if that's more amenable than this proposal, though.

Hmm, you might have a point here. The root signature approach is a bit more terse than what you're asking for it, even if it achieves something similar. 

> On a footnote, I'd like to point out that I filed the feature request since it seems like a good idea to me. It could turn out to be a bad idea and I'm open to that. But I'd like to move away from the argument that we're not a customer in discussing this. If there are technical arguments, let's bring them out first and worry about the governance problem later. I'm sympathetic to that side of the issue as well, but mixing the two tends to lead to bad results.

No one said you weren't a customer. It's just the change you're asking for is unique to the SPIR-V side, but also has the potential to be a language feature. While we always try to source ISV opinion for large changes, I understand the unique position you're in working on software at an IHV. This is the first of those requests, so we have to go through the pain of learning what that's like.

antiagainst commented:
Instead of adding `[[vk::set]]`, we propose to relax the `:register()` syntax. See #1503 for details.

antiagainst commented:
#1503 is landed. Closing this one. Thanks @pow2clk for the implementation!
